draft-ietf-tls-tls13.md

```diff
@@ -1585,28 +1582,25 @@ exist). As with ServerHello, a
 HelloRetryRequest MUST NOT contain any extensions that were not first
 offered by the client in its ClientHello.
 
-Upon receipt of a HelloRetryRequest, the client MUST first verify that
-the selected_group field corresponds to a group which was provided
-in the "supported_groups" extension in the original ClientHello.  It
-MUST then verify that the selected_group field does not correspond
-to a group which was provided in the "key_share" extension in the
-original ClientHello. If either of these checks fails, then the client
-MUST abort the handshake with a fatal "illegal_parameter"
-alert. Clients SHOULD also abort with "unexpected_message" in response
-to any second HelloRetryRequest which was sent in the same connection
-(i.e., where the ClientHello was itself in response to a
-HelloRetryRequest).
-
-Otherwise, the client MUST send a ClientHello with an updated KeyShare
-extension to the server. The client MUST append a new KeyShareEntry
-for the group indicated in the selected_group field to the groups
-in its original KeyShare.
-
-Upon re-sending the ClientHello and receiving the server's
-ServerHello/KeyShare, the client MUST verify that the selected
-NamedGroup matches that supplied in the HelloRetryRequest and MUST
-abort the connection with a fatal "illegal_parameter" alert if it does
-not.
+Upon receipt of a HelloRetryRequest, the client MUST verify that the extensions
+block is not empty and otherwise abort the handshake with a fatal
+"decode_error" alert.  Clients SHOULD also abort with "unexpected_message" in
+response to any second HelloRetryRequest which was sent in the same connection
+(i.e., where the ClientHello was itself in response to a HelloRetryRequest).
+
+Otherwise, the client MUST process all extensions in the HelloRetryRequest and
+send a second updated ClientHello. The HelloRetryRequest extensions defined in
+this specification are:
+
+- cookie [{{cookie}}]
+
+- key_share [{{key-share}}]
+
+Note that HelloRetryRequest extensions are defined such that the original
+ClientHello may be computed from the new one, given minimal state about which
+HelloRetryRequest extensions were sent. For example, the key_share extension
+causes the new KeyShareEntry to be appended to the client_shares field, rather
+than replacing it.
```

Comment:
Added special rule for `cookie` earlier where it says unexpected extensions are illegal, just to leave the semantics the same for now.

[I'm not a huge fan of `cookie` being MTI like this, to be honest. Making it negotiated doesn't cost much, though even that's not quite right. It's a function of whether the transport provides what you need here or not. TLS doesn't need it. DTLS does. I've heard competing ideas for whether QUIC should use this or do something at its lower layer, but either way it'd be "all clients MUST support" or "no need to have cookie at all". Nice thing about lower layers is you don't need to worry about ClientHello exceeding a packet, as you do in DTLS.

Anyway, we can sort all that out separately.]